python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -121,6 +121,15 @@ def _process_user_event(
         for check_result in user_event.check_results or []:
             yield from _process_user_event(step_context, check_result)
 
+        # If a MaterializeResult was returned from an asset with no type annotation, the type will be
+        # interpreted as Any and the I/O manager will be invoked. Raise a warning to alert the user.
+        if not assets_def.op.output_dict[output_name].dagster_type.is_nothing:
+            step_context.log.warning(
+                f"MaterializeResult for asset {asset_key} returned from an asset with an inferred"
+                " return type of Any. This will cause the I/O manager to run. To ensure that the"
+                " I/O manager does not run, annotate your asset with the return type MaterializeResult."
+            )
```

Comment:
I like this idea, wonder if its the right balance of useful/loud

Reply:
Yeah, we could swap to `warnings.warn` (that would just show in the process logs right?). We could also keep the compute log warning for a couple versions after 1.6 and then move it somewhere less loud.

Reply:
maybe take a couple passes at tightening up the wording of this